src/SizeBench.SKUCrawler/MasterControllerProcess.cs

```diff
@@ -1,179 +1,175 @@
+    private readonly QueuedTaskScheduler _taskScheduler = new QueuedTaskScheduler(threadCount: 3);
+    private readonly TaskFactory _taskFactory;
+    private readonly List<Task> _batchTasks = new List<Task>();
+    private readonly object _outputSyncObject = new object();
+    private readonly ConcurrentDictionary<int, ConcurrentBag<string>> _errorsFromBatchesByPID = new ConcurrentDictionary<int, ConcurrentBag<string>>();
+    private readonly ConcurrentDictionary<int, string> _batchCommandLinesByPID = new ConcurrentDictionary<int, string>();
+    private readonly ConcurrentDictionary<int, int> _batchNumbersByPID = new ConcurrentDictionary<int, int>();
+
+    public MasterControllerProcess()
+    {
+        this._taskFactory = new TaskFactory(this._taskScheduler);
+    }
+
+    public Task KickOffAndWaitForBatches(List<ProductBinary> productBinaries, CrawlFolderArguments appArgs)
+    {
+        var numBatches = (int)Math.Ceiling(productBinaries.Count / (float)appArgs.BatchSize);
+        Console.WriteLine($"Found {productBinaries.Count} binaries that SizeBench can try to parse.  Breaking this up into {numBatches} batches.");
+
+        // Note that batch number starts at 1, not zero.  Why?  I forget, but that's what it is.
+        for (var i = 1; i < numBatches + 1; i++)
+        {
+            var batchNumber = i; // Need to put this inside the loop so the lambda doesn't capture "i" as the final value over and over
+            var commandLine = appArgs.CommandLineArgsForBatch(batchNumber);
+            this._batchTasks.Add(this._taskFactory.StartNew(() => StartOneBatchTask(commandLine, batchNumber),
+                                 CancellationToken.None,
+                                 TaskCreationOptions.LongRunning,
+                                 this._taskScheduler));
+        }
+
+        return Task.WhenAll(this._batchTasks);
+    }
+
+    private void StartOneBatchTask(string commandLineForBatch, int batchNumber)
+    {
+        var psi = new ProcessStartInfo("SizeBench.SKUCrawler.exe", commandLineForBatch)
+        {
+            CreateNoWindow = true,
+            WindowStyle = ProcessWindowStyle.Hidden,
+            UseShellExecute = false,
+            RedirectStandardOutput = true,
+            RedirectStandardError = true,
+        };
+
+        Console.WriteLine($"Starting batch process, command line: {commandLineForBatch}");
+
+        var batchProcess = new Process()
+        {
+            StartInfo = psi,
+            EnableRaisingEvents = true
+        };
+        batchProcess.OutputDataReceived += BatchProcess_OutputDataReceived;
+        batchProcess.ErrorDataReceived += BatchProcess_ErrorDataReceived;
+        this.BatchProcesses.Add(batchProcess);
+        batchProcess.Start();
+        batchProcess.BeginOutputReadLine();
+        batchProcess.BeginErrorReadLine();
+        this._batchCommandLinesByPID.TryAdd(batchProcess.Id, commandLineForBatch);

```

Comment:
>TryAdd [](http://example.com/codeflow?start=37&length=6)

Try implies that this can fail.  Should it retry?

Reply:
`ConcurrentDictionary` doesn't support just `Add` - it can only try, in case it loses a race condition.

I don't think PIDs should ever be reused in practice, though, so this should be safe, it just avoids the `Dictionary` corruption that was theoretically possible before.

I don't have a local repro of this, so this is a speculative fix and if it doesn't work I'll have to try harder to get a local repro.